### PR TITLE
Plugins: Migrate Elasticsearch to backend plugin SDK

### PR DESCRIPTION
**Pull request by dsotirakis**
**What this PR does / why we need it**:

This PR is responsible for the migration of Elasticsearch to start using the backend plugin SDK

**Which issue(s) this PR fixes**:

Fixes (a checkbox in) #35179

### CODE REVIEW

**Comment by dsotirakis** [comment]
TODO: Move away from Convey!

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:96]: Would replace this with a specific typed struct for the elasticsearch data source, others like cloudwatch, opentsdb and graphite have used a struct named `datasourceInfo`. For the client package which takes `models.DataSource` as argument I would define a typed struct for that in the client package.

File: pkg/tsdb/interval/interval.go
Comment (marefr) [inline pkg/tsdb/interval/interval.go:138]: We might want to extract all/parts of the interval calculation to the SDK eventually. But as a first step might create a new calculator that operates on SDK constracts?

**Review by marefr — COMMENTED** [review]
Looks like a great start :+1: Awesome that you've converted to standard test lib :heart: 

Nit. Please remove the `// nolint:staticcheck` that doesn't apply any longer. Added some additional comments.

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/elasticsearch.go:96]: Yeah, thought about that, but also thought that it would maybe be a good practice to try and share code, since some fields in `models.DataSource` are common between datasources. WDYT?

File: pkg/tsdb/interval/interval.go
Comment (dsotirakis) [reply pkg/tsdb/interval/interval.go:138]: We are already using that on this PR, no?

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:37]: Remove?

```suggestion
```

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:38]: This seems uninitialized?

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:56]:
```suggestion
func newService(im instancemgmt.InstanceManager, cfg *setting.Cfg, httpClientProvider httpclient.Provider) *Service {
```

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:47]:
```suggestion
		QueryDataHandler: newService(im, s.Cfg, s.HTTPClientProvider),
```

**Review by marefr — COMMENTED** [review]
Nice. Added some comments/suggestions

File: pkg/tsdb/interval/interval.go
Comment (marefr) [reply pkg/tsdb/interval/interval.go:138]: I think you'll realize this is needed if you replace the usage of `models.DataSource` in the client package :) `GetIntervalFrom` is used there for example.

**Review by Elfo404 — CHANGES_REQUESTED** [review]
Finally took some time to look at this, awesome work! 🚀 

There is something that I personally consider a blocker.
Testing this out, I found out debug info is not shown anymore in the alert rule test modal: 
image compared to the previous version: 
image.

I believed this was due this change: https://github.com/grafana/grafana/pull/36132/files#diff-1e280182ca65bd6b2ece7013a14e4844d335d405f692af03bab4ca4d27a02156L45-L46 where `client.enableDebug` have been removed.

trying to add it back I also discovered that neither `context.Context` or `backend.QueryDataRequest` have anything similar to `plugins.DataQuery.Debug`.

I also enabled it without conditionals, at this point I was able to log `debugInfo` and see its content, but couldn't return the data to be displayed in that modal.

I think this is kind of a blocker as without this it would be pretty hard to understand what's going on when testing alert rules.

**Comment by marefr** [comment]
@Elfo404 good catch! Agree would also consider this a blocker for legacy alerting. Before this was really hard to investigate users having alerting problems with Elasticsearch queries. Not sure that unified alerting makes use of the debug flag because you can basically visualize the actual data response in a visualization when testing the rule.

I do believe it would be a good thing for a backend plugin to have being able to provide more detailed contextual information about a request (query) ad-hoc per request. One other thing I thought about when introducing the debug flag back in the days was to allow debug to be enabled for a data source in the configuration. On another note, I think I remember some developers raised the concern having the debug flag on the request object "feels weird".

We have 2 options as I see it:
1. Add a new field on the [QueryDataRequest protobuf message type](https://github.com/grafana/grafana-plugin-sdk-go/blob/3bbb5c542de1ac1536385d9dfe27e4ed82be4286/proto/backend.proto#L114-L122) and corresponding public API [QueryDataRequest struct](https://github.com/grafana/grafana-plugin-sdk-go/blob/3bbb5c542de1ac1536385d9dfe27e4ed82be4286/backend/data.go#L38-L42): 
2. Pass the debug flag through using context.Context (and gRPC metadata between client/server).

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:94]: Think you can use and skip the unmarshal to `map[string]interface{}`.

```suggestion
			JsonData:    simplejson.NewJson(settings.JSONData),
```

File: pkg/tsdb/elasticsearch/client/client.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/client/client.go:32]:
```suggestion
	ID             int64
	HTTPClientOpts sdkhttpclient.Options
	URL            string
	JSONData       *simplejson.Json
	Database       string
```

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:93]: Seems like Database field is not populated?

File: pkg/tsdb/elasticsearch/client/client.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/client/client.go:86]: Suggest parsing the data source configuration/JSONData in newInstanceSettings instead and put those as fields  on `DatasourceInfo`? Think code will be cleaner and good to stop using the simplejson package, [ref](https://github.com/grafana/grafana/blob/main/pkg/README.md#avoid-use-of-the-simplejson-package).

In hindsight and/or as a suggestion for the future. Think it could be a nice addition to refactor so that the whole client could be instantiated in the `newInstanceSettings` and by that making an instantiated client part of the data source instance. The only part restricting this right now is that the list of indices to be queried are embedded in the client. Don't remember right now if I had some other reason for making the client instantiated for each request rather than the lifecycle of a data source until being updated.

File: pkg/tsdb/calculator.go
Comment (marefr) [inline pkg/tsdb/calculator.go:9]: If we're considering moving this code to the SDK we don't want to reference/use the simplejson package, [ref](https://github.com/grafana/grafana/blob/main/pkg/README.md#avoid-use-of-the-simplejson-package).

Shouldn't this use backend.TimeRange rather than plugins.DataTimeRange to be compatible with the SDK?

And tests as you mentioned would be nice as well.

**Comment by Elfo404** [comment]
@marefr  I have fairly limited knowledge of the BE architecture, but I believe adding a new field in QueryDataRequest makes more sense (this is just a sensation, my opinion here is for sure less valuable than yours).

I also agree that a "debug" flag might feel weird, but regarding
>Not sure that unified alerting makes use of the debug flag because you can basically visualize the actual data response in a visualization when testing the rule.

although it might not be used at the moment, I still think it could be a really valuable thing to add as it would be quite hard (especially for non-devs) to look at the queries running under the hood (this is probably a different topic, but even if it's not going to be displayed in the UI, it would be much easier to look at the HTTP response - where I assume this info is going to be present - rather than starting a debug session in grafana server)

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:34]: Don't think this is in use by this package?

```suggestion
```

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:38]: Typo? Duplicate?

```suggestion
```

File: pkg/tsdb/elasticsearch/time_series_query.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/time_series_query.go:87]: Shouldn't this use the new `tsdb.IntervalCalculator`?

File: pkg/tsdb/testdatasource/testdata/flight-simple-X.txt
Comment (marefr) [inline pkg/tsdb/testdatasource/testdata/flight-simple-X.txt:3]: Why did this change?

**Review by marefr — CHANGES_REQUESTED** [review]
Added some comments/suggestions. Looks good in general, but think there's opportunity to simplify the code even more and remove usage of simplejson package.

File: pkg/tsdb/interval/interval.go
Comment (marefr) [reply pkg/tsdb/interval/interval.go:138]: Added some comments in the new file. Guess the changes in this file can be reverted then?

**Comment by marefr** [comment]
@Elfo404 @dsotirakis talked with Ryan and I basically remembered that CloudWatch are using a similar flag to denote a query is originating from Grafana alerting. CloudWatch are checking the request `Headers` for a certain property, see https://github.com/grafana/grafana/blob/17707d886adae2e2fc08ede607c0dace62f6c63f/pkg/tsdb/cloudwatch/cloudwatch.go#L277

So for the short term solution I would add a header if Debug field is true somehwere in here https://github.com/grafana/grafana/blob/92cb50348685e9a8d29ba1257b9b9b1425683a30/pkg/tsdb/data_plugin_adapter.go#L23

WDYT?

File: pkg/tsdb/testdatasource/testdata/flight-simple-X.txt
Comment (dsotirakis) [reply pkg/tsdb/testdatasource/testdata/flight-simple-X.txt:3]: That's related to a PR merged by Ryan, was needed back when I started my PR, will not be needed when I merge with main anymore :)

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/elasticsearch.go:93]: Yeah, I think it's not needed anymore.

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/elasticsearch.go:34]: Probably not needed at all, yes.

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/elasticsearch.go:38]: Probably not needed at all, yes.

**Review by Elfo404 — APPROVED** [review]

**Comment by Elfo404** [comment]
Approving to unblock this as we discussed offline 🚀

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:96]: Should check that these casts is allowed/ok. 

Regarding `esVersion` not sure that would handle version stored as either string or integer. Either need to move the version parsing that currently happens in client.go to to here and use a `*semver.Version` on the `DatasourceInfo ESVersion` field or either try cast to both string and integer.

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:89]: Looks like you're missing populating the `MaxConcurrentShardRequests` field?

File: pkg/tsdb/elasticsearch/time_series_query.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/time_series_query.go:401]: Don't think this is correct. The value here is string representation of the interval, e.g. `"10s"`. Need to apply something like this 
https://github.com/grafana/grafana/blob/2026a1dfb6a013f255d465fbac2aed1f83564268/pkg/tsdb/interval/interval.go#L92-L103

**Review by marefr — CHANGES_REQUESTED** [review]
Unfortunately there's still a few things left before we can merge this. Maybe you can ask @idafurjes to incorporate the `pkg/tsdb/calculator.go` in her PR for now to unblock her?

File: pkg/tsdb/elasticsearch/time_series_query.go
Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/time_series_query.go:401]: Hm, I think I just wrongly c/p-ed

`intervalMs := model.Get("interval").MustInt64(0)`, needed to be: `intervalMs := model.Get("intervalMs").MustInt64(0)`

and also, I don't know if `intervalMs` is string, since I can see that:
```
if interval == "" {
		intervalMS := queryModel.Get("intervalMs").MustInt(0)
		if intervalMS != 0 {
			return time.Duration(intervalMS) * time.Millisecond, nil
		}
	}
```

Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/time_series_query.go:401]: In addition, the code you mentioned above, is still used in `calculator.go`

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/elasticsearch.go:96]: `esVersion` is being checked in `coerceVersion(v string)`?

Comment (marefr) [reply pkg/tsdb/elasticsearch/elasticsearch.go:96]: Was referring to something like this in `newInstanceSettings` since `jsonData["esVersion"]` could be either a string or an integer. 

```go
esVersion, err := coerceVersion(jsonData["esVersion"])
if err != nil {
	return nil, err
}

model.ESVersion = esVersion

func coerceVersion(v interface{}) (*semver.Version, error) {
	versionString, err := v.String()
	if err == nil {
		return semver.NewVersion(versionString)
	}

	versionNumber, err := v.Int()
	if err != nil {
		return nil, err
	}

	// Legacy version numbers (before Grafana 8)
	// valid values were 2,5,56,60,70
	switch versionNumber {
	case 2:
		return semver.NewVersion("2.0.0")
	case 5:
		return semver.NewVersion("5.0.0")
	case 56:
		return semver.NewVersion("5.6.0")
	case 60:
		return semver.NewVersion("6.0.0")
	case 70:
		return semver.NewVersion("7.0.0")
	default:
		return nil, fmt.Errorf("elasticsearch version=%d is not supported", versionNumber)
	}
}
```

File: pkg/tsdb/elasticsearch/time_series_query.go
Comment (marefr) [reply pkg/tsdb/elasticsearch/time_series_query.go:401]: I was referring to if you skip introducing `intervalMs := model.Get("interval").MustInt64(0)` and use the existing interval string representation (`interval := model.Get("interval").MustString("")`). 

Then change the `tsdb.GetIntervalFrom` to 
```go
func GetIntervalFrom(interval, timeInterval string, queryInterval string, defaultInterval time.Duration) (time.Duration, error)
```

And `GetMinInterval` to 
```go
GetMinInterval(queryInterval string) (time.Duration, error)
```

> In addition, the code you mentioned above, is still used in calculator.go

Might have missed that part, great. But seems like both interval string and intervalMs integer is pass around now?

I might be confused/misunderstood something :smile: Let me know if my suggestions make sense?

Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/time_series_query.go:401]: `intervalMs` is actually `int64`, since it's milliseconds, and it's different than `interval` which is `string`. So if the plan is to move away from `simpleJson` we need to declare it explicitly I think. Let's discuss more offline.

Comment (marefr) [reply pkg/tsdb/elasticsearch/time_series_query.go:401]: Sorry I'm super confused now. Yes, lets discuss it.

> So if the plan is to move away from simpleJson we need to declare it explicitly I think

My whole point is to explicitly use a string rather than an int64 (as you've seem to advise).

For queries originating from old alerting system the property "interval" of string is part of the query model. To remove usage of simplejson here you should probable decode the JSON into `Query` using `json.Unmarshal`.

File: pkg/tsdb/calculator.go
Comment (marefr) [inline pkg/tsdb/calculator.go:102]: Changed the order of the arguments to make this API more clear to the end-user. As discussed, no longer want to move this to the SDK so maybe moving this into the `interval` package to keep all logic on one place could make sense, but up to you when you want to do that.

```suggestion
// GetIntervalFrom returns the minimum interval. 
// dsInterval is the string representation of data source min interval, if configured.
// queryInterval is the string representation of query interval (min interval), e.g. "10ms" or "10s".
// queryIntervalMS is a pre-calculated numeric representation of the query interval in milliseconds.
func GetIntervalFrom(dsInterval, queryInterval string, queryIntervalMS int64, defaultInterval time.Duration) (time.Duration, error) {
	if queryInterval == "" {
		if queryIntervalMS != 0 {
			return time.Duration(queryIntervalMS) * time.Millisecond, nil
		}
	}

	interval := queryInterval
	if queryInterval == "" && dsInterval != "" {
		interval = dsInterval
		
	}

	if interval == "" {
		return defaultInterval, nil
	}

	interval = strings.Replace(strings.Replace(interval, "<", "", 1), ">", "", 1)
	isPureNum, err := regexp.MatchString(`^\d+$`, interval)
	if err != nil {
		return time.Duration(0), err
	}
	if isPureNum {
		interval += "s"
	}
	parsedInterval, err := time.ParseDuration(interval)
	if err != nil {
		return time.Duration(0), err
	}

	return parsedInterval, nil
}
```

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/elasticsearch.go:130]: What's the difference between these? Think only `timeInterval` is stored in jsonData.

File: pkg/tsdb/elasticsearch/time_series_query.go
Comment (marefr) [inline pkg/tsdb/elasticsearch/time_series_query.go:401]: As discussed, this isn't needed/applicable for old alerting.

```suggestion
```

**Review by marefr — COMMENTED** [review]
After our discussions added some comments/suggestions. Think we can get this merged after some of these are addressed.

File: pkg/tsdb/calculator_test.go
Comment (marefr) [inline pkg/tsdb/calculator_test.go:19]:
```suggestion
```

**Review by marefr — COMMENTED** [review]
Tests looks good.

File: pkg/tsdb/elasticsearch/elasticsearch.go
Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/elasticsearch.go:130]: Seen both, that's why I kept them!

Comment (marefr) [reply pkg/tsdb/elasticsearch/elasticsearch.go:130]: Hmm okay. Sure you haven't seen `timeInterval` on the data source config and `interval` on each query?

Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/elasticsearch.go:130]: Yeah, I could see both on `JSONData`.

Have a look here for `interval`: https://github.com/grafana/grafana/pull/36132/files#diff-95c1bcdd942efc587e6f08b4175e52e30410f94425bc990a9260fb446f131831L90

Comment (marefr) [reply pkg/tsdb/elasticsearch/elasticsearch.go:130]: Oops you're absolutely correct. Forgot about that one. Unfortunate it's being named `interval` as well :joy: Great

**Review by marefr — APPROVED** [review]
LGTM :tada: great work, haven't tested it though but expects that you others have

**Pull request merged**
